### PR TITLE
Exoforce

### PR DESCRIPTION
**Pull request by MarisaRipoll**

### CODE REVIEW

File: src/exoforce.py [hunk 1]
```diff
@@ -1,3 +1,10 @@
+#######################################################################################
```
Comment (missxa) [inline src/exoforce.py:1]: add .gitignore
do not push __pycache__ folder

File: src/exoforce.py [hunk 2]
```diff
@@ -16,10 +23,42 @@
...
+30: Right Leg (Point where it joins the waist)
+31: Right Knee
+33: Right Foot
+38: Left Leh (Point where it joins the waist)
+39: Left Knee
+41: Left Foot
+"""
+
+# LINK DEFINITIONS
+righthand_link = 16
```
Comment (missxa) [inline src/exoforce.py]: find link by string, do not hardcode ids

File: src/exoforce.py [hunk 3]
```diff
@@ -28,24 +67,25 @@
 for i in range(joints):
 	info = p.getJointInfo(human, i)
 	print(info)
-info = p.getJointInfo(human, hand_link)
+info = p.getJointInfo(human, lefthand_link)
 print(info)
 
-# NUMBER OF CONSTRAINTS = 0 (Unless I uncomments createConstraint function)
+# NUMBER OF CONSTRAINTS = 0 (Initially - later constraints can be set)
```
Comment (missxa) [inline src/exoforce.py]: what is this for? never used

File: src/exoforce.py [hunk 4]
```diff
@@ -28,24 +67,25 @@
...
 print("The number of Constraints is: ", cons)
 
 for i in range(cons):
 	info = p.getConstraintInfo(i)
 	print(info)
 
-# GET LINK STATES, JOINT STATES
-linkstate = p.getLinkState(human, hand_link)
-print("The link state or link hand_link is: ", linkstate)
+# GET LINK AND JOINT STATES
```
Comment (missxa) [inline src/exoforce.py]: unused

File: src/exoforce.py [hunk 5]
```diff
@@ -28,24 +67,25 @@
...
+linkstate = p.getLinkState(human, lefthand_link)
+print("The link state or link lefthand_link is: ", linkstate)
 print("\n")
 
-jointinfo = p.getJointInfo(human, hand_link)
-print("The joint info or link hand_link is: ", jointinfo)
+jointinfo = p.getJointInfo(human, lefthand_link)
```
Comment (missxa) [inline src/exoforce.py]: unused

File: src/exoforce.py [hunk 6]
```diff
@@ -28,24 +67,25 @@
...
+print("The joint info or link lefthand_link is: ", jointinfo)
+print("\n")
 
 # NUMBER OF BODIES = 2 (0 = plane, 1 = human)
 bodies = p.getNumBodies()
```
Comment (missxa) [inline src/exoforce.py]: unused

File: src/exoforce.py [hunk 7]
```diff
@@ -57,103 +97,132 @@
 	print("This body's Id is: ", body_id)"""
 
 
-# COMMON SETUP
+# COMMON WORLD SETUP
 p.setGravity(0, 0, -9.81)
-# p.setTimeStep(-0.0001)
-p.setRealTimeSimulation(0) # XXX
+p.setRealTimeSimulation(0)  # Requirement for p.applyExternalForce() to work     # XXX
 color = [1, 0, 0]
```
Comment (missxa) [inline src/exoforce.py]: color defined again

File: src/exoforce.py [hunk 8]
```diff
@@ -57,103 +97,132 @@
...
+###########    DEBUGLINES, PARAMS, CONSTRAINTS AND FORCE DEFINITIONS     #############
+###########                                                              #############
+######################################################################################
+
+
+# LINK POSITION DEFINITIONS:
+lefthand_pos = p.getLinkState(human, lefthand_link)[0]
+righthand_pos = p.getLinkState(human, righthand_link)[0]
+
+# TENDON ORIGIN DEFINITIONS:
```
Comment (missxa) [inline src/exoforce.py]: remove copy-paste and implement a Tendon class and move all relevant functions there 
```python
class Tendon:
       self.start_location = xx
       self.end_location =xx
       self.current_force = xx

       def __init__(self, start_location, end_location):
           ...

       def decompose_force(self, force):
           ...

       def apply_force(self, force):
           ...
       
       def visualize(self):
          ...
 
       def update(self, start, end, force):
          self.decompose_force()
          self.apply_force()
          self.visualize()
 
```

Comment (missxa) [reply src/exoforce.py]: the structure is proposed for guidance, you'll likely need additional params and member functions

File: src/exoforce.py [hunk 9]
```diff
@@ -57,103 +97,132 @@
...
-p.createConstraint(human,		# ParentBodyUniqueId
-		   22,			# ParentLinkIndex
-		   hand_link,			# childBodyUniqueId
-		   hand_link,			# childLinkIndex
-		   4,			# Join Type: JOINT_GEAR
-		   [0,0,1],		# Joint Axis
-		   [0,0,0],		# ParentFramePosition
-		   [0,0,0])		# childFramePosition"""
-
-# THIS IS THE BIGGEST ISSUE, THE FIGURE DOES NOT REACT TO THE FUNCTION APPLYEXTERNALFORCE
```
Comment (missxa) [inline src/exoforce.py:125]: remove the comments like this

File: src/exoforce.py [hunk 10]
```diff
@@ -57,103 +97,132 @@
...
-# 		     flags=2)		# 2 Flags: WORLD_FRAME and LINK_FRAME
-
-# THE CODE LINE BELOW RECEIVES THE SAME REACTION AS THE CREATECONSTRAINT FUNCTION:
-# p.setJointMotorControlArray(human, (21,22,hand_link), p.POSITION_CONTROL)
-hand_pos = p.getLinkState(human, hand_link)[0]
-tendon_origin = [hand_pos[0]+0.5, hand_pos[1]+0.5, 3.0]
-# motor_force = 50
-forceId = p.addUserDebugParameter("motor_force", 0,200,0)
 
 for _ in range(100000000):
```
Comment (missxa) [inline src/exoforce.py]: what is this magic number 100000000

File: src/simulation_for_video.py [hunk 11]
```diff
@@ -0,0 +1,309 @@
+import argparse
```
Comment (missxa) [inline src/simulation_for_video.py]: this file should not exist, it's basically a copy of exoforce.py

**Review by missxa — CHANGES_REQUESTED** [review]
remove unnecessary code and comments

File: src/exoforce.py [hunk 12]
```diff
@@ -20,209 +19,116 @@
...
-		   -1,
-		   3,
-		   p.JOINT_FIXED,
-		   [0, 0, 0],
-		   [0, 0, 0],
-		   [0, 0, 1])	"""	
+p.setRealTimeSimulation(0)     # Don't change. Else p.applyExternalForce() won't work.
 
+# TENDONS:
+t1 = Tendon(human, lefthand_link, [0.5, 0.5, 1.0], 1)
```
Comment (missxa) [inline src/exoforce.py]: store tendons in the list

File: src/exoforce.py [hunk 13]
```diff
@@ -20,209 +19,116 @@
...
 ######################################################################################
 ###########                                                              #############
-###########    ACTUAL SIMULATION ---> DEBUGLINES AND EXTERNAL FORCES     #############
+###########      ACTUAL SIMULATION (UPDATING DEBUGLINES AND FORCES)      #############
 ###########                                                              #############
 ######################################################################################
 
+try:
+	while True:
+		t1.update()
```
Comment (missxa) [inline src/exoforce.py]: loop the tendon list here

**Review by missxa — CHANGES_REQUESTED** [review]
looking ok for now,

address minor comments, add o the readme instruction how to run the exoforce script, and we're good to go

File: src/exoforce.py [hunk 14]
```diff
@@ -13,147 +19,116 @@
...
+#######################################################################################
+
+
+# LINK DEFINITIONS
+hf = humanFigure(human)
+
+lefthand_link = hf.find_link_number(b'human/left_wrist')
+righthand_link = hf.find_link_number(b'human/right_wrist')
+
+upperchest_link = hf.find_link_number(b'human/shoulder_center')
```
Comment (missxa) [inline src/exoforce.py]: also unclear why do you define all these link ids that are never used

Comment (MarisaRipoll) [reply src/exoforce.py]: When we are done with the optimization of key link points we will probably add more tendons on other links and so, I thought that preparing them might be helpful moving forward. This doesn't apply to legs though. I will delete those straight away. Should I also delete the other unused ones?

**Comment by MarisaRipoll** [comment]
By address minor comments you mean I should add more minor comments right?

**Review by missxa — APPROVED** [review]

**Pull request merged**